core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala

```diff
@@ -313,6 +313,21 @@ class GroupCoordinatorTest {
     assertEquals(Errors.INCONSISTENT_GROUP_PROTOCOL, joinGroupResult.error)
   }
 
+  @Test
+  def testCompleteNewMemberTimeoutHeartbeat(): Unit = {
+    // New member joins the group
+    var responseFuture = sendJoinGroup(groupId, JoinGroupRequest.UNKNOWN_MEMBER_ID, protocolType, protocols, None, DefaultSessionTimeout, DefaultRebalanceTimeout, false)
+    timer.advanceClock(GroupInitialRebalanceDelay + 1)
+
+    val joinResult = Await.result(responseFuture, Duration(DefaultRebalanceTimeout + 100, TimeUnit.MILLISECONDS))
+    val group = groupCoordinator.groupManager.getGroup(groupId).get
+    assertEquals(Errors.NONE, joinResult.error)
+    assertEquals(0, group.allMemberMetadata.count(_.isNew))
+
+    // Make sure the delayed heartbeat based on the new member timeout has been completed
+    assertEquals(1, groupCoordinator.heartbeatPurgatory.numDelayed)
```

Comment:
If we wanted to go a tad further, we could advance the clock by the session timeout and assert the member gets kicked out. That ensures that the heartbeat left behind has the right timeout.